### PR TITLE
r106-acropolis

### PR DESCRIPTION
**Pull request by knservis**
TODO:

- [x] Smoke test
- [x] Blog post (in release branch, passed QA)
- [ ] Blog post (in master)
- [x] Wiki in release branch passed QA
- [ ] Wiki in master

### CODE REVIEW

**Comment by knservis** [comment]
@chuwy @BenFradet Smoke testing went well (including with the rc1 of shredder). We have looked through all this code in previous PRs. Let me know if you spot something new. Thanks

File: 3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/registry/pii/PiiPseudonymizerEnrichment.scala [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+    case s: String =>
+      val newValue = strategy.scramble(s)
+      val _        = modifiedFields += JsonModifiedField(fieldName, s, newValue, jsonPath, schema)
+      newValue
+    case a: ArrayNode =>
+      a.elements.asScala.map {
+        case t: TextNode =>
+          val originalValue = t.asText()
+          val newValue      = strategy.scramble(originalValue)
+          val _             = modifiedFields += JsonModifiedField(fieldName, originalValue, newValue, jsonPath, schema)
```
Comment (BenFradet) [inline 3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/registry/pii/PiiPseudonymizerEnrichment.scala]: why is this a val?

Comment (knservis) [reply 3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/registry/pii/PiiPseudonymizerEnrichment.scala]: Early version leftover

**Comment by knservis** [comment]
@BenFradet Thanks not sure what is missing from the milestone.

**Comment by knservis** [comment]
@chuwy Thanks

**Comment by knservis** [comment]
@BenFradet Can you please look at the last commit before I copy it to all sources?

**Comment by BenFradet** [comment]
last commit looks good :+1:

**Comment by knservis** [comment]
@BenFradet Thanks. Same approach for all the rest. (I will soon rebase)

File: 3-enrich/stream-enrich/nsq/src/main/scala/com.snowplowanalytics.snowplow.enrich.stream/sources/NsqSource.scala [hunk 2]
```diff
@@ -67,15 +96,20 @@ class NsqSource private (
...
-    override def initialValue: Sink = new NsqSink(nsqConfig, config.out.enriched)
+    override def initialValue: Sink = new NsqSink(goodProducer, config.out.enriched)
   }
 
-  override val threadLocalPiiSink: Option[ThreadLocal[Sink]] = if(emitPii(enrichmentRegistry)) Some(new ThreadLocal[Sink] {
-    override def initialValue: Sink = new NsqSink(nsqConfig, config.out.pii)
-  }) else None
+  override val threadLocalPiiSink: Option[ThreadLocal[Sink]] = piiProducer.flatMap {
+    somePiiProducer =>
+      config.out.pii.map { piiTopicName =>
```
[the remark this answers is not in the document]
Comment (knservis) [reply 3-enrich/stream-enrich/nsq/src/main/scala/com.snowplowanalytics.snowplow.enrich.stream/sources/NsqSource.scala]: So what would you do instead?

[the remark this answers is not in the document]
Comment (BenFradet) [reply 3-enrich/stream-enrich/nsq/src/main/scala/com.snowplowanalytics.snowplow.enrich.stream/sources/NsqSource.scala]: nvm I thought https://github.com/snowplow/snowplow/pull/3766/files/81d0b8248a4fcfee0d5438262b63c61b91d169fe..230cb00266b95fbac7a7eaa511b8c35d9af60d36#diff-b9d72942ccb9985afcce223303adf230R35 was used

[the remark this answers is not in the document]
Comment (BenFradet) [reply 3-enrich/stream-enrich/nsq/src/main/scala/com.snowplowanalytics.snowplow.enrich.stream/sources/NsqSource.scala]: I wouldn't call when creating it either

[the remark this answers is not in the document]
Comment (knservis) [reply 3-enrich/stream-enrich/nsq/src/main/scala/com.snowplowanalytics.snowplow.enrich.stream/sources/NsqSource.scala]: Not quite sure what you mean. You mean creating the producer on NsqSource instantiation?

[the remark this answers is not in the document]
Comment (BenFradet) [reply 3-enrich/stream-enrich/nsq/src/main/scala/com.snowplowanalytics.snowplow.enrich.stream/sources/NsqSource.scala]: the word __start__ was missing from the above sorry, what I meant was:

```scala
object NsqSink {
  def validateAndCreateProducer(nsqConfig: Nsq): \/[Throwable, NSQProducer] =
    new NSQProducer().addAddress(nsqConfig.host, nsqConfig.port).right
}

class NsqSink(...) {
  producer.start()
}
```

**Comment by knservis** [comment]
@BenFradet I think we are also done with the restructuring.. Anything else left?

**Comment by BenFradet** [comment]
:+1: from me

**Pull request merged**